### PR TITLE
Export type annotations for public use

### PR DESCRIPTION
**Pull request by bgilbert**
Type annotations are currently internal to pyfuse3, so applications can't use them for their own type checking.  Fix some bugs and omissions in the annotations and make them public by adding a `py.typed` file.

To do so, we're [required](https://peps.python.org/pep-0561/#packaging-type-information) to move our modules into a Python package.  Do this, but add forwarders for the `_pyfuse3` and `pyfuse3_asyncio` modules to avoid breaking compatibility.

Probably best reviewed as individual commits.  GitHub's rendering of the overall diff is somewhat unhelpful.

### CODE REVIEW

File: src/pyfuse3/_pyfuse3.py [hunk 1]
```diff
@@ -403,7 +404,7 @@ async def read(
         fh: FileHandleT,
         off: int,
         size: int
-    ) -> int:
+    ) -> bytes:
```
Comment (ThomasWaldmann) [inline src/pyfuse3/_pyfuse3.py]: oops, good catch!

File: src/pyfuse3/__init__.pyi [hunk 2]
```diff
@@ -9,7 +9,15 @@ This file is part of pyfuse3. This work may be distributed under
...
-from ._pyfuse3 import Operations, async_wrapper, FileHandleT, FileNameT, FlagT, InodeT, ModeT
+# re-exports
+from ._pyfuse3 import (
+    Operations as Operations,
+    FileHandleT as FileHandleT,
+    FileNameT as FileNameT,
+    FlagT as FlagT,
+    InodeT as InodeT,
+    ModeT as ModeT
+)
```
Comment (ThomasWaldmann) [inline src/pyfuse3/__init__.pyi:20]: Yet I only used "from ... import ... as ..." when needing to rename.
So just importing these here is not sufficient?

**Review by ThomasWaldmann — COMMENTED** [review]
Had a quick look, lgtm - but I am not that familiar with type annotations and cython.

**Comment by bgilbert** [comment]
Okay, this should be ready for review.  CI should pass now.  This update fixes some additional type issues, adds a new `XAttrNameT` for xattr names (subclassing `bytes`), and adds a mypy run to CI.

I've gone through [pyfuse3-stubs](https://pypi.org/project/pyfuse3-stubs/) and compared its types to the ones in pyfuse3, incorporating any fixes I found.  pyfuse3-stubs includes additional types for some pyfuse3 internals like `g` and `_NANOS_PER_SEC`, which are omitted here.

cc:
- @odgalvin &mdash; original contributor of pyfuse3 types
- @rhelmot &mdash; author of pyfuse3-stubs, which I think is superseded by this PR
- @Nikratio &mdash; cc'd in an edit of a comment above, which IIRC doesn't generate a notification

**Comment by bgilbert** [comment]
Updated once again, to fix docs and to re-export the NewTypes at runtime for use e.g. with `typing.cast()`.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -40,15 +40,19 @@ jobs:
         if: ${{ matrix.cython-version == '3.0' }}
         run: pip install "Cython>=3"
 
-      - name: Test
+      - name: Build
```
Comment (ThomasWaldmann) [inline .github/workflows/test.yml:43]: Good catch!

**Comment by bgilbert** [comment]
@Nikratio, per https://github.com/libfuse/pyfuse3/pull/95#pullrequestreview-1744612706, is there any chance you could take a look?

**Comment by Nikratio** [comment]
Sorry for the delay. It seems that `src/pyfuse3/_pyfuse3.py` is not recognized as a rename but instead a completely new file, which makes the diff very large. Do you know why this might be the case? Did you perhaps do large whitespace changes by accident?

**Comment by bgilbert** [comment]
It's probably because there's still a `src/_pyfuse3.py`, it's just a stub now.  I'd recommend reviewing commit-by-commit.  I tried to keep each commit legible and self-contained.

**Comment by bgilbert** [comment]
Thanks for the reviews!

**Comment by bgilbert** [comment]
Do you have a sense of when this might end up in a release?

**Comment by ThomasWaldmann** [comment]
@bgilbert there is only 1 small issue left to do in the 3.3.1 milestone, so guess I could do a release in the near future.

**Pull request merged**